icicle_v3/backend/cpu/include/cpu_ntt.h

```diff
@@ -0,0 +1,331 @@
+#include "icicle/ntt.h"
+#include "icicle/errors.h"
+#include "icicle/runtime.h"
+#include "icicle/utils/log.h"
+#include "icicle/fields/field_config.h"
+
+#include <vector>
+#include <algorithm>
+#include <iostream>
+#include <cmath>
+#include <cstdint>
+#include <memory>
+#include <mutex>
+
+using namespace field_config;
+using namespace icicle;
+
+namespace cpu_ntt_h {
+
+  template <typename S>
+  class CpuNttDomain
+  {
+    int max_size = 0;
+    int max_log_size = 0;
+    std::unique_ptr<S[]> twiddles;
+    std::mutex domain_mutex;
+
+  public:
+    static eIcicleError
+    cpu_ntt_init_domain(const Device& device, const S& primitive_root, const NTTInitDomainConfig& config);
+
+    static eIcicleError cpu_ntt_release_domain(const Device& device);
+
+    template <typename U, typename E>
+    eIcicleError
+    cpu_ntt_ref(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output);
+
+    template <typename U, typename E>
+    eIcicleError cpu_ntt(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output);
+
+    const S* get_twiddles() const { return twiddles.get(); }
+    const int get_max_size() const { return max_size; }
+
+    static inline CpuNttDomain<S> s_ntt_domain;
+  };
+
+  template <typename S>
+  eIcicleError
+  CpuNttDomain<S>::cpu_ntt_init_domain(const Device& device, const S& primitive_root, const NTTInitDomainConfig& config)
+  {
+    // (1) check if need to refresh domain. TODO SHANIE - reusing domain for different ntt sizes if possible
+    if (s_ntt_domain.twiddles != nullptr) { return eIcicleError::SUCCESS; }
+
+    // Lock the mutex to ensure thread safety during initialization
+    std::lock_guard<std::mutex> lock(s_ntt_domain.domain_mutex); // Added for thread safety
+
+    // Double-checked locking
+    if (s_ntt_domain.twiddles == nullptr) {
+      // (2) build the domain
+
+      bool found_logn = false;
+      S omega = primitive_root;
+      const unsigned omegas_count = S::get_omegas_count();
+      for (int i = 0; i < omegas_count; i++) {
+        omega = S::sqr(omega);
+        if (!found_logn) {
+          ++s_ntt_domain.max_log_size;
+          found_logn = omega == S::one();
+          if (found_logn) break;
+        }
+      }
+
+      s_ntt_domain.max_size = (int)pow(2, s_ntt_domain.max_log_size);
+      if (omega != S::one()) {
+        ICICLE_LOG_ERROR << "Primitive root provided to the InitDomain function is not in the subgroup";
+        return eIcicleError::INVALID_ARGUMENT;
+      }
+
+      // calculate twiddles
+      // Note: radix-2 INTT needs ONE in last element (in addition to first element), therefore have n+1 elements
+
+      std::unique_ptr<S[]> temp_twiddles(new S[s_ntt_domain.max_size + 1]);
```

Comment:
please add comment why you need the temp twiddles. It's because otherwise twiddles are not nullptr but still not computed right?

BTW you'd better use std::make_unique<S[]>(max_size+1) instead of new.